samples/nrf9160/nrf_cloud_mqtt_multi_service/README.rst

```diff
@@ -0,0 +1,297 @@
+
+* Connection Thread (``conn_thread``): Runs the `Connection Management Loop <nrf_cloud_mqtt_multi_services_sample_connection_management_loop_>`_, which maintains a connection to the `nRF Cloud MQTT API`_.
+* Application Thread (``app_thread``): Runs the `Main Application Loop <nrf_cloud_mqtt_multi_services_sample_main_application_loop_>`_, which may add `Device Messages <nRF Cloud Device Messages_>`_ to the `Device Message Queue <nrf_cloud_mqtt_multi_services_sample_device_message_queue_>`_.
+* Message Thread (``msg_thread``): Processes the `Device Message Queue <nrf_cloud_mqtt_multi_services_sample_device_message_queue_>`_ whenever there is an active connection.
+
+.. _nrf_cloud_mqtt_multi_services_sample_connection_management_loop:
+
+Connection Management Loop
+==========================
+
+The Connection Management Loop performs four primary tasks:
+
+1) Configure and activate the cellular modem.
+#) Start a connection to the LTE Network.
+#) Maintain a connection to the `nRF Cloud MQTT API`_.
+#) Reestablish or retry all lost or failed connections.
+
+Modem setup and the LTE connection are set-and-forget.
+If connection to the LTE is lost, the modem automatically works to reestablish it.
+
+The connection to the `nRF Cloud MQTT API`_, on the other hand, must be maintained manually by the application.
+A significant portion of :file:`src/connection.c` is dedicated to this task.
+
+This file also maintains a number of `Zephyr Events`_ tracking the state of the lte and `nRF Cloud MQTT API`_ connections.
+The Connection Management Loop relies on these events to control its progression.
+
+The Connection Management Loop must conform to the :ref:`nRF Cloud connection process <lib_nrf_cloud_connect>`.
+This is mostly handled by the :ref:`nrf_cloud library<lib_nrf_cloud>`, with exception to one task:
+Whenever a user association request is received from `nRF Cloud`_, the application must manually disconnect and reconnect.
+One of the cases inside the :func:`cloud_event_handler` function handles this.
+
+The application must also manually reset its connection to `nRF Cloud`_ whenever connection to the LTE lost, or else the :ref:`nrf_cloud library<lib_nrf_cloud>` will deadlock.
+One of the cases inside the :func:`lte_event_handler` function performs this task.
+
+.. _nrf_cloud_mqtt_multi_services_sample_device_message_queue:
+
+Device Message Queue
+====================
+
+Any thread may submit `Device Messages <nRF Cloud Device Messages_>`_ to the Device Message Queue, where they will be stored until a working connection to the `nRF Cloud MQTT API`_ is established.
+Once this happens, the Message Thread transmits all enqueued `Device Messages <nRF Cloud Device Messages_>`_, one at a time and in fast succession, to `nRF Cloud`_.
+Transmission pauses again whenever connection to the `nRF Cloud MQTT API`_ is lost.
+Enqueued messages may be lost whenever the device transitions from connected to not connected.
```

Comment:
The transition itself, mainly because the device wrongly thinks it is still connected and attempts to transmit the message, and has no backup/recourse for if transmission fails.

Reply:
I wonder if MQTT QoS=1 would help?

Also, does it have a chance to try again if there is an immediate error returned from whatever function does the sending?

Reply:
I'm not sure, for some disconnection types QoS=1 might help, for others it might not. I can look into it.

As for trying again, I did not implement a re-enqueue-on-failure. I had some reservations regarding sample complexity and retry limits, but in retrospect, I don't expect the outgoing message itself to ever be the issue, so indefinite retry shouldn't be a problem, and if I stick to just indefinite retry, then the added complexity is minimal. So I'll go ahead and implement that. Message transmission order will not be guaranteed, but I don't think that was ever guaranteed with MQTT anyways.

I'll want to bring back the minimum delay between message sends, though, @jayteemo, because otherwise the sample could potentially spin indefinitely trying to resend the same messages over and over again.

Reply:
This is done now